File: scratch/test_url_transform.js

function transformGoogleUrl(url) {
    if (!url) return url;
    
    const driveMatch = url.match(/\/file\/d\/([^\/?#]+)/);
    if (driveMatch) {
        return `https://drive.google.com/uc?export=download&id=${driveMatch[1]}`;
    }

    const slideMatch = url.match(/\/presentation\/d\/([^\/?#]+)/);
    if (slideMatch) {
        return `https://docs.google.com/presentation/d/${slideMatch[1]}/export/pptx`;
    }

    const docMatch = url.match(/\/document\/d\/([^\/?#]+)/);
    if (docMatch) {
        return `https://docs.google.com/document/d/${docMatch[1]}/export?format=docx`;
    }

    const sheetMatch = url.match(/\/spreadsheets\/d\/([^\/?#]+)/);
    if (sheetMatch) {
        return `https://docs.google.com/spreadsheets/d/${sheetMatch[1]}/export?format=xlsx`;
    }

    return url;
}

const testUrls = [
    "https://drive.google.com/file/d/1QPefrkTlIcOqq_syqZhw4Opgjs2hrJjC/view?usp=classroom_web",
    "https://docs.google.com/presentation/d/1abcd1234/edit#slide=id.p",
    "https://docs.google.com/document/d/1doc_id_here/edit?tab=t.0",
    "https://docs.google.com/spreadsheets/d/1sheet_id/edit",
    "https://example.com/somefile.pptx"
];

testUrls.forEach(url => {
    console.log(`Original: ${url}`);
    console.log(`Transformed: ${transformGoogleUrl(url)}`);
    console.log('---');
});

File: scratch/test_url_transform.py

import re

def transform_google_url(url):
    if not url: return url
    
    # Extract account index if present (e.g. /u/0/ or /u/1/)
    auth_match = re.search(r'\/u\/([0-9]+)\/', url)
    auth_part = f"/u/{auth_match.group(1)}" if auth_match else ""
    
    # Extract ID (usually follows /d/)
    id_match = re.search(r'\/d\/([^\/?#]+)', url)
    if not id_match: return url
    file_id = id_match.group(1)

    # 1. Google Drive Files - Using docs.google.com
    if 'drive.google.com/file' in url:
        return f"https://docs.google.com{auth_part}/uc?export=download&id={file_id}"

    return url

test_urls = [
    "https://drive.google.com/file/u/1/d/1QPefrkTlIcOqq_syqZhw4Opgjs2hrJjC/view"
]

for url in test_urls:
    print(f"Original: {url}")
    print(f"Transformed: {transform_google_url(url)}")
    print('---')
